internal/js/modules/k6/browser/common/frame_manager.go

```diff
@@ -540,7 +540,40 @@ func (m *FrameManager) requestStarted(req *Request) {
 		frame.pendingDocument = &DocumentInfo{documentID: req.documentID, request: req}
 		frame.pendingDocumentMu.Unlock()
 	}
-	m.logger.Debugf("FrameManager:requestStarted", "fmid:%d rurl:%s pdoc:nil", m.ID(), req.URL())
+
+	if !m.page.hasRoutes() {
+		return
+	}
+
+	route := NewRoute(m.logger, m.page.mainFrameSession.networkManager, req)
+	for _, r := range m.page.routes {
+		matcher, err := urlMatcher(r.path, r.jsRegexChecker)
+		if err != nil {
+			m.logger.Errorf("FrameManager:requestStarted",
+				"fmid:%d rurl:%s error creating url matcher: %v", m.ID(), req.URL(), err)
+			continue
+		}
+
+		matched, err := matcher(req.URL())
+		if err != nil {
+			m.logger.Errorf("FrameManager:requestStarted",
+				"fmid:%d rurl:%s error matching url: %v", m.ID(), req.URL(), err)
+			continue
+		}
+
+		if !matched {
+			continue
+		}
+
+		err = r.handler(route)
```

Comment:
Could we create a new instance of `route` just before passing it to `handler`?

Reply:
I'm not sure it makes sense. I would understand if it avoids creating it and not using it but here the route object will be used every time, either here or in the `route.Continue` line at the end. 
If you think the route creation and its usage is too far await, the update in https://github.com/grafana/k6/pull/4953/commits/48787ab2a9c41c510eed541f1fafba49167c4ecb improves this.